src/Lecture2.hs

```diff
@@ -157,14 +180,83 @@ You're free to define any helper functions.
        treasure besides gold (if you already haven't done this).
 -}
 
--- some help in the beginning ;)
-data Knight = Knight
-    { knightHealth    :: Int
-    , knightAttack    :: Int
+-- -- some help in the beginning ;)
+type Health = Int 
+type Attack = Int
+type Gold = Int
+
+data Knight = MKKnight
+    { knightHealth    :: Health
+    , knightAttack    :: Attack
     , knightEndurance :: Int
     }
+    
+data Chest a = MKChest 
+    { chestGold    :: Gold
+    , chestTresure :: a
+    }
+
+data DragonType 
+    = Red
+    | Black
+    | Green
+
+data Dragon = MKDragon 
+    { dragonType      :: DragonType
+    , dragonHealth    :: Health
+    , dragonFirePower :: Attack
+    }
+
+data Cave a = MKCave 
+    { caveChest      :: Chest a
+    , caveDragon     :: Dragon
+    }
+
+data Reward a = MKReward 
+    { rewardTreasure   :: Maybe a
+    , rewardFold       :: Gold
+    , rewardExpirience :: Int
+    }
 
-dragonFight = error "TODO"
+data FightResult a
+    = Win (Reward a)
+    | Run
+    | Death
+
+attack :: Int -> Knight -> Dragon -> (Int, Knight, Dragon)
+attack attempt knight dragon
+    | attempt == 10 = 
+        ( 0
+        , knight { knightHealth = knightHealth knight - dragonFirePower dragon }
+        , dragon
+        )
+    | otherwise = 
+        ( attempt + 1
+        , knight { knightEndurance = knightEndurance knight - 1 }
+        , dragon { dragonHealth = dragonHealth dragon - knightAttack knight }
+        )
+
+-- Need to pass chest for polymorphism, how can I fix it? 🤔 
+resultOrNextAttack :: ((Int, Knight, Dragon), Chest a) -> FightResult a
```

Comment:
I'm not sure I understand the question 🤔 
The implementation looks correct 🙂 
Could you clarify what do you mean?

Reply:
This question is not relevant anymore, cleared it for myself later 🙂